### PR TITLE
Use gravity fix

### PR DESCRIPTION
**Pull request by eirikh1996**
### Describe in detail what your pull request acomplishes
In this pull request, the functionalities of useGravity, canHover, canHoverOverWater and hoverLimit craft flags are re-implemented

### Related issues:
- #75 useGravity not working

### Checklist
- [x] Unit tests 
- [x] Proper internationalization
- [x] Compiled/tested

### CODE REVIEW

**Comment by eirikh1996** [comment]
@cccm5 Requested changes made

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/craft/ICraft.java [hunk 1]
```diff
@@ -96,56 +96,66 @@ public void rotate(Rotation rotation, MovecraftLocation originPoint) {
...
             }
+            new BukkitRunnable(){
+                @Override
+                public void run() {
+                    switch (move){
+                        case DOWN:
+                            translate(0, -1, 0);
+                            break;
+                        case UP:
+                            Movecraft.getInstance().getAsyncManager().submitTask(new RotationTask(craft, originPoint, rotation, craft.getW()), craft);
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/craft/ICraft.java]: This will not move the craft up

Comment (eirikh1996) [reply modules/Movecraft/src/main/java/net/countercraft/movecraft/craft/ICraft.java]: It is being moved up before the BukkitRunnable is run

**Comment by eirikh1996** [comment]
So @cccm5 I have done some improvements in the code where gravity is being processed on rotation, which includes, but not limited to, the value of "move" being processed in a single switch statement. When it comes to the order the translation and rotation is being processed, on up, it has to first translate the craft up, then rotate it to avoid it being obstructed on rotate. On down, however, that order has to be in reverse to avoid the craft being obstructed when being translated. Trust me, I did test the code when the order was the same in both situations, and it worked in one situation, but not the other

**Comment by cccm5** [comment]
The issue is *where* you are doing the rotation checks. This should not be happening in the ICraft class, but instead one of the business logic handlers for rotation. When you try to translate the craft, they both happen on the same tick because you bypass the queue by doing it when rotation is initially scheduled.

**Comment by eirikh1996** [comment]
@cccm5 Requested changes made

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/async/translation/TranslationTask.java [hunk 2]
```diff
@@ -90,6 +87,24 @@ protected void execute() {
...
+            } else if (!isOnGround(oldHitBox) && craft.getType().getCanHover()){
+                MovecraftLocation midPoint = oldHitBox.getMidPoint();
+                int centreMinY = oldHitBox.getLocalMinY(midPoint.getX(), midPoint.getZ());
+                int groundY = centreMinY;
+                World w = craft.getW();
+                while (w.getBlockAt(midPoint.getX(), groundY - 1, midPoint.getZ()).getType() == Material.AIR || craft.getType().getPassthroughBlocks().contains(w.getBlockAt(midPoint.getX(), groundY - 1, midPoint.getZ()).getType())){
+                    groundY--;
+                }
+                if (centreMinY - groundY > craft.getType().getHoverLimit()){
+                    dy = -1;
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/async/translation/TranslationTask.java]: be consistent in the use of `--` and `-=` - preferably the former.

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/async/rotation/RotationTask.java [hunk 3]
```diff
@@ -195,6 +207,65 @@ protected void execute() {
...
+            craft.rotate(rotation, originPoint, false);
+            new BukkitRunnable(){
+                @Override
+                public void run() {
+                    craft.translate(0, -1, 0);
+
+                }
+            }.runTaskLaterAsynchronously(Movecraft.getInstance(), 2);
+            return;
+        }
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/async/rotation/RotationTask.java]: The only thing different between what is processed in moveup and movedown is the vector at which the craft translates. There's no need to repeat all of the code twice.

Comment (eirikh1996) [reply modules/Movecraft/src/main/java/net/countercraft/movecraft/async/rotation/RotationTask.java]: The order of when rotate and translate is called is also different

**Comment by eirikh1996** [comment]
@cccm5 Requested changed made

File: modules/api/src/main/java/net/countercraft/movecraft/craft/Craft.java [hunk 4]
```diff
@@ -53,6 +53,7 @@
     private long lastCruiseUpdate;
     private long lastBlockCheck;
     private long lastRotateTime=0;
+    private long lastGravityOnRotateTime=0;
```
Comment (cccm5) [inline modules/api/src/main/java/net/countercraft/movecraft/craft/Craft.java]: This field is never modified/referenced

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/mapUpdater/update/CraftTranslateCommand.java [hunk 5]
```diff
@@ -13,19 +13,30 @@
...
     @NotNull private final MovecraftLocation displacement;
+    @Nullable private final HashHitBox newHitBox;
 
     public CraftTranslateCommand(@NotNull Craft craft, @NotNull MovecraftLocation displacement){
         this.craft = craft;
         this.displacement = displacement;
+        this.newHitBox = null;
     }
 
+    public CraftTranslateCommand(@NotNull Craft craft, @NotNull MovecraftLocation displacement, @NotNull HashHitBox newHitBox){
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/mapUpdater/update/CraftTranslateCommand.java]: This constructor is never used

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/mapUpdater/update/CraftTranslateCommand.java [hunk 6]
```diff
@@ -13,19 +13,30 @@
...
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.util.*;
 import java.util.logging.Logger;
 
 public class CraftTranslateCommand extends UpdateCommand {
     @NotNull private final Craft craft;
     @NotNull private final MovecraftLocation displacement;
+    @Nullable private final HashHitBox newHitBox;
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/mapUpdater/update/CraftTranslateCommand.java]: This field is never used

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/mapUpdater/update/CraftTranslateCommand.java [hunk 7]
```diff
@@ -168,7 +179,10 @@ public void doUpdate() {
                 }
             }
         }
-
+        //Update hitbox if new hitbox is not null
+        if (newHitBox != null){
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/mapUpdater/update/CraftTranslateCommand.java]: always null

File: modules/Movecraft/src/main/java/net/countercraft/movecraft/async/translation/TranslationTask.java [hunk 8]
```diff
@@ -412,6 +440,75 @@ private boolean checkFuel(){
...
+            }
+        }
+        boolean translatedBottomLocsInAir = true;
+        for (MovecraftLocation translatedBottomLoc : translatedBottomLocs){
+            MovecraftLocation beneath = translatedBottomLoc.translate(0, -1, 0);
+            Material testType = beneath.toBukkit(craft.getW()).getBlock().getType();
+            if (bottomLocs.contains(beneath) || testType == Material.AIR || craft.getType().getPassthroughBlocks().contains(testType)){
+                continue;
+            }
+            translatedBottomLocsInAir = false;
```
Comment (cccm5) [inline modules/Movecraft/src/main/java/net/countercraft/movecraft/async/translation/TranslationTask.java]: break the loop here

**Comment by eirikh1996** [comment]
Requested changes made @cccm5 . Everything is tested and working

**Pull request merged**